### DISCUSSION TITLE
delta filter - behaviour (just checking)

### DISCUSSION THREAD

**Discussion by luissimoes** [dbuezas/lovelace-plotly-graph-card#377] [Q&A] [answered]
I have some inconsistent results when using a statistic of type sum and then using a filter: delta.

Just to be sure:
the documentation says: _delta # computes the delta between each two consecutive numeric y values_.

**Does it provide also negative values to the sum?** or just the positive values

**Comment by dbuezas** [comment]
It just the previous value from the last.
https://github.com/dbuezas/lovelace-plotly-graph-card/blob/master/src/filters/filters.ts#L111

Can you share your yaml and a screenshot of both the original statistic (without filters) and with it?

**Comment by luissimoes** [comment]
Here is what I have.... I dont understand what is going on, with the measurement (all details hereafter).


**Situation**
I measure the _solar_energy_consumed_ through a sensor built on a template: I subtract the to_netz (energy sold) from the solar_energy_produced.

To generate the graph I use statistic: sum, period: day, filter: delta
```
  - entity: sensor.solar_energy_consumed
    texttemplate: '%{y:.2f}'
    textposition: outside
    textfont:
      size: 12
    name: solar consumed
    type: bar
    statistic: sum
    period: day
    marker:
      color: '#97cd07'
      opacity: 0.9
    filters:
      - delta
      - filter: i>0 & i < xs.length - 1
      - fn: |
          ({ys,xs,hass, vars}) => ({
            xs: [...xs, +xs.at(-1)+1000*60*60*24],
            ys: [...ys, hass.states['sensor.solar_energy_consumed'].state],
          })
```


The result is in the figure (includes the separate graph for solar energy produced)
<img width="979" alt="Pasted Graphic 15" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/70d96b70-ca59-400e-8c57-aebc6b0f4d88">


**Problems**
There are clearly wrong values for the solar_energy_consumed in the graph, although the individual values of the sensor are correct

> 2024-02-14: should be 3.39 -  is 7.08
> 2024-02-16: should be 2.32 - is: 6.16

<img width="700" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/4dcaad8a-ef2f-4f61-bbf5-41764173c771">

--> a possible problem is off the "peaks" .... **but that is why I dont understand how the stats work i.e. count**


**On the peaks** 
They are created by 2 reasons:  (pictures at the end to improve readability)
-the ones during the day is because the information sent by the sensors sometimes is not available; I defaulted it by making the value 0 (zero) if it has no value
- the ones at 00:00 is because the sensor for to_netz resets exactly at 00:00 and the sensor for solar_energy_produce a few seconds (10?) later.... **so the solar_energy_consumed *at the start of the new day* is still the same as at 00:00**


**My Questions and (not-)Conclusions** 
this was the reason for my post: **how is the delta calculated?**
- if the delta counts also "negative" increments (which is apparently does) then the peaks in the middle of the day **have very little influence** because the state before and after the stabilisation is the same (the ups are compensated by the downs and results in a net zero change)

the behaviours at 00:00 are puzzling
- the day starts with a value V that then goes (after 10 seconds) to 0..... **so there is a -V delta** that should be added by the statistics
- if the stats decide that (because the period is day) the day starts with zero, then there is a delta of V (the persistent value) and a delta of -V... and so the net value is zero....
- in this case **there should also be no impact in the final value... ---> but the values are wrong and impacted!** and it seems indeed that the reason is this peak at 00:00....



_________________________
Add-Ons

**_template for solar_energy_consumed_**
```
{% set solar_prod = (states('sensor.solar_energy_daily_yield') |float(0)/1000) %}
{% set energy_sold = states('sensor.to_netz_daily')|float(0) %}
{% set solar_consumed = (solar_prod - energy_sold) | float(0) %}

{% if solar_consumed <= 0 %}
  {{ 0 }}
{% else %}
  {{ solar_consumed }}
{% endif %}
```

**Pictures for the details of the peaks**

1) middle of the day
As can be seen the sensor is not available and as a result the sensor used in the template has a value of zero, and so the the solar_energy_consumed is zero...
Later the sensor for the to_netz is briefly not available and as a result (it is read as zero) the solar_energy_consumed is counted as everything the was produced....

<img width="367" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/27754840-b0b3-4483-95e6-8a541a74b879">

2) midnight
Here the interesting case
The sensor to_netz resets at 00:00 and the other seconds later... so the solar_energy_consumed *at the start of the new day* is still the same as at 00:00

<img width="568" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/b574f058-4c6c-4e6c-8603-bb8efd76501d">

**Reply by dbuezas** [reply]
Can you make a plot with all input traces and the result too? 
And make it a scatter plot so it is easier to compare.

Watch out with the single ampersand &, that's the bitwise and and has different presence than the logical and &&.

Regarding the delta filter, it just computes `ys[i]-ys[i-1]`

**Reply by luissimoes** [reply]
Hi, 
here is the graph, but I dont think this explains or shows the problem completly....  
(have then please a look at the details I am showing below this)

![image](https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/e7091939-be20-4081-b4fd-67f9ecd5ab1d)

- in fact it shows there is a problem (or I am not understanding how delta works... 
-> I know how it works but I think the result is not correct --- delta, being the difference between current and previous **should also account for negative values... and seems it doesn't**

Here is why - let's pick the period 2024-02-18-23:55  to 2024-02-19-00:05. (so the midnight)

1) The sensor solar_energy_daily_yield **holds** the value after midnight.... (doesn't reset immediately) (where as the to_netz_daily does)
2) as a result: solar_energy_consumed also holds the top value.
3) then it goes to zero.... ---> this means delta should have a negative value of -5 which should be reflected in the total of the day
... but it doesn't

---> unless ... the delta starts also at zero at 00:00 and then it has a +5 and an -5 i.e. doesn't look at the Current value

![image](https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/72e3b93c-d965-4d98-b032-e2c2e9d051de)


![image](https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/98d7b954-6be3-4f22-8a7e-139d117ffb70)



Thanks!


_________________
The sensors:
- solar_energy_consumed = solar_energy_daily_yield - to_netz_daily
```
{% set solar_prod = (states('sensor.solar_energy_daily_yield') |float(0)/1000) %}
{% set energy_sold = states('sensor.to_netz_daily')|float(0) %}
{% set solar_consumed = (solar_prod - energy_sold) | float(0) %}

{% if solar_consumed <= 0 %}
  {{ 0 }}
{% else %}
  {{ solar_consumed }}
{% endif %}
```

Solar_energy_daily_yield (in orange) is the same as solar_panels_daily_yeld ... except that it delivers zero when the latter is not available (which happens)... 
```
{% if has_value('sensor.pv_dailyyield') %}
  {{ states('sensor.pv_dailyyield') }}
{% else %}
  0
{% endif %}
```

**Reply by dbuezas** [reply]
Please post your yaml

**Reply by luissimoes** [reply]
Hi,
Here is the graphs I prepared: I have 2 sensors:
- sensor.pv_dailyyield: comes directly from the photovoltaic installation, and sometimes is the sensor that is unavailable)
- sensor.solar_daily_energy: its from the utility.yaml... and counts the energy from the above sensor daily i.e. it resets at 00:00


My conclusion for now:
"delta": resets at midnight i.e. doesn't count the value that is "brought" over at the time
Meaning:
- if the value at 00:00:00 was eg 5 and then at 00:05 is 0, I would have expected "delta" to show a value of -5
- from the example it seems that it counts from 0 to 5 (+5) and then from 5 to 0 (-5) resulting in a net zero value.

**Note: for me this is OK and perfect**. But the behaviour was not clear and I wanted to clear because it can affect calculations a lot.


The data

YAML
```
type: custom:plotly-graph
entities:
  - entity: sensor.pv_dailyyield
    type: bar
    statistic: sum
    period: day
    yaxis: y1
    marker:
      color: cyan
      opacity: 0.8
    filters:
      - delta
      - map_y_numbers: y/1000
  - entity: sensor.solar_daily_energy
    type: bar
    statistic: sum
    period: day
    yaxis: y1
    marker:
      color: orange
      opacity: 0.8
    filters:
      - delta
      - map_y_numbers: y/1000
hours_to_show: 72
refresh_interval: 10
title: All inputs
```

Now lets get the graph for the 2 sensors.
**As can be seen the totals are exactly the same** - no difference in adding.

<img width="472" alt="Solar Panels Daily yield" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/90903926-97f5-4694-bd7f-8f5b4b15bb2c">

Let's look at the details at midnight (00:00) for instance for the 18. to 19. (it's the same whatever the day).

---> **the sensor.pv_dailyyield has a "holdover" onto the next day.**.. 

==== Delta Behaviour
1) if delta doesn't reset then the first change is a Negative value (around 2.8) and so the values of the graph could never be the same as shown
2) the values **are** the same... so this means that in this statistic, the delta "resets" at 00:00 and then as a positive and a negative increment, totalling net zero


______


For the sensor.pv_dailyyield
<img width="956" alt="Pasted Graphic 53" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/18ec7ff1-3867-4b81-94f8-784393626604">


For the sensor.solar_daily_energy
<img width="956" alt="Pasted Graphic 54" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/43f5bf41-5c50-4559-940f-7743288150ae">

**Reply by dbuezas** [reply]
I think the issue is that these are sum daily sensors. We only get 1 datapoint per day.
Sorry I can't go really deep analysing the problem at the moment, but I suggest plot the raw entities (without the delta filter) too. This way you can see the input to the delta and debug this better.
To deal with the reset issue, you could try `period: 5minute` and use a resample filter. This way you get more datapoints per day and can do something about the sensor resetting too early and breaking the total.

**Reply by luissimoes** [reply]
Thanks for the answers and helping

Actually the sensors update a few times per minute - while there is sun :-) - and they are energy increasing so they always have a value.

I solved using state and a sensor that resets at 00:00 defined in the utility yaml.

My point was to understand the delta behavior and seems it resets at 00:00 - that’s why there isn’t a negative delta in the final sum

**Reply by dbuezas** [reply, marked as answer]
The delta filter just subtracts the previous from each datapoint. Are you sure you are not confusing statistics with history?
Statistics is a different db in home assistant, and when you ask for daily statistics of a sensor, you only get 1 datapoint per day. If that statistic is configured to reset each day then that happens way before this gets into this card.

Try plotting this:
```yaml
type: custom:plotly-graph
entities:
  - entity: sensor.pv_dailyyield # history db, not statistic db. Typically only holds one week of data.
  - entity: sensor.pv_dailyyield # statistics db, holds data indefinitely 
    statistic: sum
    period: day
  - entity: sensor.pv_dailyyield
    statistic: sum
    period: hour
  - entity: sensor.pv_dailyyield  # also statistics db but granlar, isn't kept forever
    statistic: sum
    period: 5minute
```


This will show you exactly what data comes in.

**Reply by luissimoes** [reply]
AH! that's clear now!!! I was not aware that one would get *one* value for the period ... makes a lot of sense!
And the plot you proposed is amazing to clarify! That you!
I plotted it and also added sum and delta... just to see it....

<img width="481" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/fe538b21-2741-4d1b-ae8f-a4f6a7e31487">

I see now .... (there is light at the end of the tunnel) :-)

THANKS!
All clear now!

what has always made me confused was the behaviour of the sensor that **holds** the value 20 seconds or so after midnight....
(see below - this is the same pv_dailyyield)
<img width="646" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/38a329a6-ff61-4087-817e-e841e5c88141">

**Reply by dbuezas** [reply]
🎉 Great! At first I didn't understand what your issue was and then it clicked :) 
Happy to hear this helped.

Regarding the late update, it may well be that the sensor updates every once a while, and some seconds after midnight just happens to be the first update of the day.

**Reply by dbuezas** [reply]
BTW, if you want to use the daily statistic but also have a bar with the current partial data of today, see this https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/299#discussioncomment-6527714

**Reply by luissimoes** [reply]
thanks!
I am using now statistic: state, period: day (since it brings one point as you mentioned and it is the real value at the end of the day)... so its perfect!

I got a graph like the one below.. which is what I wanted to show all the components.

But regarding your suggestion....
your suggestion is (for the first filter)

Alternative 1)
```
    filters:
      - filter: i>0 && i < xs.length - 1
      - fn: |
          ({ys,xs,hass}) => ({
            xs: [...xs, new Date().setHours(0,0,0)],
            ys: [...ys, hass.states['sensor.from_netz_daily'].state],
          })

```

and I had (Alternative 2)
```
    filters:
      - filter: i>0 && i < xs.length - 1
      - fn: |
          ({ys,xs,hass, vars}) => ({
            xs: [...xs, +xs.at(-1)+1000*60*60*24],
            ys: [...ys, hass.states['sensor.from_netz_daily'].state],
          })
```
I guess it was from the tentatives to have the graph center values since I am getting the last value and adding a day

**===> what is best? the first of the second? from a system point of view?**
or is it the same?

thanks


<img width="501" alt="image" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/44951cd3-cd46-4592-944e-3fe98be62f78">

**Reply by dbuezas** [reply]
Wow, that's looking very cool!
Would you make a show&tell post with the image and yaml? I bet somebody else is trying to do the same thing somewhere in the world :) 

> ===> what is best? the first of the second? from a system point of view?
> or is it the same?

They're almost equivalent. I would say the first one may be better because it will still show today's data on the right date (today), even if the statistic happens to be missing yesterday for whatever reason.

**Reply by luissimoes** [reply]
thanks :-). ... but you are responsible for it too :-)

Ok... if this can help someone I will put it in show&tell. Thanks for the tip!

Luis
